### ISSUE TITLE
Odd behavior with site.sun_set_time

### ISSUE THREAD

**Issue by gnarayan** [astropy/astroplan#261]
I've been using astroplan (v0.2) to schedule observations for the LCO global network, in particular for their 1m telescopes at sso, saao, ctio and mcdonald sites. In the process, I've found one very confusing bug:

`
import astroplan as ap
site = ap.observer.Observer.at_site('mcdonald')
date = '2016-12-30'
from astropy.time import Time

site.sun_rise_time(Time(date,format='isot'), which='next')
<Time object: scale='utc' format='jd' value=2457753.0816>

site.sun_set_time(Time(date,format='isot'), which='next')
WARNING: TargetNeverUpWarning: Target(s) index [0] does not cross horizon=0.0 deg within 24 hours [astroplan.observer]
<Time object: scale='utc' format='jd' value=-999.0>
`

This doesn't happen for any of the other sites, and of course, the negative JD causes PriorityScheduler and plot_airmass to have very odd behavior, should you try to build a Schedule with it.

**Comment by gnarayan** [comment]
I think I do understand why this is happening. It's when the provided time is really close to sunset itself. Adjusting the time a little bit (.01 day) is sufficient to get the right value, but it'd be nice to have somewhat more predictable behavior here.

`
site.sun_set_time(Time(2457752.5,format='jd'), which='next')
<Time object: scale='utc' format='jd' value=-999.0>

 site.sun_set_time(Time(2457752.49,format='jd'), which='next')
<Time object: scale='utc' format='jd' value=2457752.49987>

site.sun_set_time(Time(2457752.51,format='jd'), which='next')
<Time object: scale='utc' format='jd' value=2457753.50035>
`

**Comment by bmorris3** [comment]
This has been a point of confusion before, and we should find a better way to handle it, and probably add another note to the docs like this [note on precision](http://astroplan.readthedocs.io/en/latest/faq/precision.html).

Currently, rise/set times are calculated by looking for the roots in the altitude of a target over the previous/next 24 hours. This can lead to bugs like this one. For example, if a user asks when the _next_ sunrise is relative to a time immediately after sunrise, there may be _no_ sunrises in the next 24 hours, if the time of year being tested is in the autumn, when the sun rises later each day. That's why the error that gets thrown is that the target is "never up", because it hasn't risen within 24 hours.

One possible (but maybe problematic) fix could be to change the duration of the root search from 24 hours to 24.1 hours, which would never miss the Sun. But hen there can be multiple rises/sets within that time span, and we'd need to be careful to return the correct one. 

cc @eteq @ejeschke @cdeil

**Comment by gnarayan** [comment]
My 2 cents is that it is better returning None in this case, rather than a valid Time object with an invalid value... 
For now, I'm just asking for 'next' sunset and if I get JD -999.0, then get 'nearest'. This is the "right" thing in our case. The 24.1 hour fix would be problematic, because there are different objects to be scheduled for the next night.

**Comment by bmorris3** [comment]
Originally, the warning that you're seeing raised was an error. Some of the developers thought that raising an error was too harsh, since a user may try to do scheduling computations on targets that never rise/set accidentally, and the whole scheduling script should not crash as a result.

We originally wanted to use `JD = nan` times as outputs in this scenario, but it turns out that `astropy.time.Time` doesn't support `nan`s (astropy/astropy4032), and it probably won't any time soon. Lacking `nan`s, we decided on having the `JD=-999` "magic value" back in this [discussion](https://github.com/astropy/astroplan/pull/34#issuecomment-128089175), so that we could do searches for this error time, which is unlikely to arise naturally from any user inputs, and we can interpret these times as "not useful times".

Could you describe in more detail how the 24.1 hour fix for the sunrise/sunset time wouldn't work for you?

**Comment by gnarayan** [comment]
Sure. Given a date, and site, we want to find sunset, and the next sunrise to that sunset to construct the Schedule object for that site. For the LCO observations, we've pre-computed nights following a geometric spacing through the semester to probe variability, and simply want to find the best windows on a given night and site for all the requested blocks. There are four separate sites that we have to schedule for, but each block can have only one request, and must include the requested windows for the same observation from all four sites. 

We're trying to not break the geometric spacing, and getting the "next" night through a 24.1 hour fix breaks this pattern. That alone might not make much difference, but we're also trying to get simultaneous spectroscopy with SOAR, and so one site giving a <date> while the other giving us <date>+1 day is tricky (especially if this happens with SOAR, since what dates we're assigned isn't under our control). astroplan has made most of this a breeze, but I did inadvertently run into the bug I reported here because of the -999.0 JD. 

Here's a visual of how densely spaced this gets:
[2017-03-19_schedule.pdf](https://github.com/astropy/astroplan/files/567363/2017-03-19_schedule.pdf)
